// x402-solana/dist/utils/index.mjs
// src/types/x402-protocol.ts
var SOLANA_MAINNET_CAIP2 = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
var SOLANA_DEVNET_CAIP2 = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1";
function isSolanaMainnet(network) {
  return network === "solana" || network === SOLANA_MAINNET_CAIP2;
}

// src/utils/helpers.ts
var USDC_MAINNET = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
var USDC_DEVNET = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";
function createPaymentPayload(transaction, paymentRequirements, resourceUrl) {
  const base64Transaction = Buffer.from(transaction.serialize()).toString(
    "base64"
  );
  const paymentPayload = {
    x402Version: 2,
    resource: {
      url: resourceUrl,
      description: paymentRequirements.extra?.description || "",
      mimeType: paymentRequirements.extra?.mimeType || "application/json"
    },
    accepted: paymentRequirements,
    payload: {
      transaction: base64Transaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString("base64");
  return paymentHeader;
}
function createPaymentPayloadV1(transaction, paymentRequirements) {
  const base64Transaction = Buffer.from(transaction.serialize()).toString("base64");
  const paymentPayload = {
    x402Version: 1,
    scheme: paymentRequirements.scheme,
    network: paymentRequirements.network,
    payload: {
      transaction: base64Transaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString(
    "base64"
  );
  return paymentHeader;
}
function getDefaultRpcUrl(network) {
  if (isSolanaMainnet(network)) {
    return "https://api.mainnet-beta.solana.com";
  }
  return "https://api.devnet.solana.com";
}
function getRpcUrlForNetwork(network) {
  if (network === SOLANA_MAINNET_CAIP2) {
    return "https://api.mainnet-beta.solana.com";
  }
  if (network === SOLANA_DEVNET_CAIP2) {
    return "https://api.devnet.solana.com";
  }
  return "https://api.devnet.solana.com";
}
function getDefaultTokenAsset(network) {
  if (isSolanaMainnet(network)) {
    return {
      address: USDC_MAINNET,
      decimals: 6
    };
  }
  return {
    address: USDC_DEVNET,
    decimals: 6
  };
}
function toAtomicUnits(amount, decimals) {
  return Math.floor(amount * Math.pow(10, decimals)).toString();
}
function fromAtomicUnits(atomicUnits, decimals) {
  return Number(atomicUnits) / Math.pow(10, decimals);
}

export { createPaymentPayload, createPaymentPayloadV1, fromAtomicUnits, getDefaultRpcUrl, getDefaultTokenAsset, getRpcUrlForNetwork, toAtomicUnits };
