// x402-solana/dist/index.mjs
import { safeBase64Decode } from '@payai/x402/utils';
import { Connection, PublicKey, ComputeBudgetProgram, TransactionMessage, VersionedTransaction } from '@solana/web3.js';
import { TOKEN_2022_PROGRAM_ID, TOKEN_PROGRAM_ID, getMint, getAssociatedTokenAddress, createTransferCheckedInstruction } from '@solana/spl-token';
import { getOrGenerateJwt } from '@payai/facilitator';

// src/types/x402-protocol.ts
var SOLANA_MAINNET_CAIP2 = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
var SOLANA_DEVNET_CAIP2 = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1";
function isSolanaNetwork(network) {
  return network === "solana" || network === "solana-devnet" || network.startsWith("solana:");
}
function isSolanaMainnet(network) {
  return network === "solana" || network === SOLANA_MAINNET_CAIP2;
}
function isSolanaDevnet(network) {
  return network === "solana-devnet" || network === SOLANA_DEVNET_CAIP2;
}
function toCAIP2Network(network) {
  switch (network) {
    case "solana":
      return SOLANA_MAINNET_CAIP2;
    case "solana-devnet":
      return SOLANA_DEVNET_CAIP2;
  }
}
function toSimpleNetwork(network) {
  if (network === SOLANA_MAINNET_CAIP2 || network === "solana") {
    return "solana";
  }
  return "solana-devnet";
}

// src/utils/helpers.ts
var USDC_MAINNET = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
var USDC_DEVNET = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";
function createPaymentPayload(transaction, paymentRequirements, resourceUrl) {
  const base64Transaction = Buffer.from(transaction.serialize()).toString(
    "base64"
  );
  const paymentPayload = {
    x402Version: 2,
    resource: {
      url: resourceUrl,
      description: paymentRequirements.extra?.description || "",
      mimeType: paymentRequirements.extra?.mimeType || "application/json"
    },
    accepted: paymentRequirements,
    payload: {
      transaction: base64Transaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString("base64");
  return paymentHeader;
}
function createPaymentPayloadV1(transaction, paymentRequirements) {
  const base64Transaction = Buffer.from(transaction.serialize()).toString("base64");
  const paymentPayload = {
    x402Version: 1,
    scheme: paymentRequirements.scheme,
    network: paymentRequirements.network,
    payload: {
      transaction: base64Transaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString(
    "base64"
  );
  return paymentHeader;
}
function getDefaultRpcUrl(network) {
  if (isSolanaMainnet(network)) {
    return "https://api.mainnet-beta.solana.com";
  }
  return "https://api.devnet.solana.com";
}
function getRpcUrlForNetwork(network) {
  if (network === SOLANA_MAINNET_CAIP2) {
    return "https://api.mainnet-beta.solana.com";
  }
  if (network === SOLANA_DEVNET_CAIP2) {
    return "https://api.devnet.solana.com";
  }
  return "https://api.devnet.solana.com";
}
function getDefaultTokenAsset(network) {
  if (isSolanaMainnet(network)) {
    return {
      address: USDC_MAINNET,
      decimals: 6
    };
  }
  return {
    address: USDC_DEVNET,
    decimals: 6
  };
}
function toAtomicUnits(amount, decimals) {
  return Math.floor(amount * Math.pow(10, decimals)).toString();
}
function fromAtomicUnits(atomicUnits, decimals) {
  return Number(atomicUnits) / Math.pow(10, decimals);
}
var DEFAULT_COMPUTE_UNIT_LIMIT = 2e4;
var DEFAULT_COMPUTE_UNIT_PRICE_MICROLAMPORTS = 1;
async function createSolanaPaymentTransaction(wallet, paymentRequirements, rpcUrl) {
  const connection = new Connection(rpcUrl, "confirmed");
  const feePayer = paymentRequirements.extra?.feePayer;
  if (!feePayer) {
    throw new Error(
      "Missing facilitator feePayer in payment requirements (extra.feePayer)."
    );
  }
  const feePayerPubkey = new PublicKey(feePayer);
  const walletAddress = wallet?.publicKey?.toString() || wallet?.address;
  if (!walletAddress) {
    throw new Error("Missing connected Solana wallet address or publicKey");
  }
  const userPubkey = new PublicKey(walletAddress);
  if (!paymentRequirements.payTo) {
    throw new Error("Missing payTo in payment requirements");
  }
  const destination = new PublicKey(paymentRequirements.payTo);
  if (!paymentRequirements.asset) {
    throw new Error("Missing token mint for SPL transfer");
  }
  const mintPubkey = new PublicKey(paymentRequirements.asset);
  const instructions = [];
  instructions.push(
    ComputeBudgetProgram.setComputeUnitLimit({
      units: DEFAULT_COMPUTE_UNIT_LIMIT
    })
  );
  instructions.push(
    ComputeBudgetProgram.setComputeUnitPrice({
      microLamports: DEFAULT_COMPUTE_UNIT_PRICE_MICROLAMPORTS
    })
  );
  const mintInfo = await connection.getAccountInfo(mintPubkey, "confirmed");
  const programId = mintInfo?.owner?.toBase58() === TOKEN_2022_PROGRAM_ID.toBase58() ? TOKEN_2022_PROGRAM_ID : TOKEN_PROGRAM_ID;
  const mint = await getMint(connection, mintPubkey, void 0, programId);
  const sourceAta = await getAssociatedTokenAddress(
    mintPubkey,
    userPubkey,
    false,
    programId
  );
  const destinationAta = await getAssociatedTokenAddress(
    mintPubkey,
    destination,
    false,
    programId
  );
  const sourceAtaInfo = await connection.getAccountInfo(sourceAta, "confirmed");
  if (!sourceAtaInfo) {
    throw new Error(
      `User does not have an Associated Token Account for ${paymentRequirements.asset}. Please create one first or ensure you have the required token.`
    );
  }
  const destAtaInfo = await connection.getAccountInfo(
    destinationAta,
    "confirmed"
  );
  if (!destAtaInfo) {
    throw new Error(
      `Destination does not have an Associated Token Account for ${paymentRequirements.asset}. The receiver must create their token account before receiving payments.`
    );
  }
  const amountStr = paymentRequirements.amount || paymentRequirements.maxAmountRequired;
  if (!amountStr) {
    throw new Error("Missing amount in payment requirements");
  }
  const amount = BigInt(amountStr);
  instructions.push(
    createTransferCheckedInstruction(
      sourceAta,
      mintPubkey,
      destinationAta,
      userPubkey,
      amount,
      mint.decimals,
      [],
      programId
    )
  );
  const { blockhash } = await connection.getLatestBlockhash("confirmed");
  const message = new TransactionMessage({
    payerKey: feePayerPubkey,
    recentBlockhash: blockhash,
    instructions
  }).compileToV0Message();
  const transaction = new VersionedTransaction(message);
  if (typeof wallet?.signTransaction !== "function") {
    throw new Error("Connected wallet does not support signTransaction");
  }
  const userSignedTx = await wallet.signTransaction(transaction);
  return userSignedTx;
}

// src/client/payment-interceptor.ts
function decodePaymentRequiredHeader(header) {
  const decoded = safeBase64Decode(header);
  return JSON.parse(decoded);
}
function createPaymentFetch(fetchFn, wallet, rpcUrl, maxValue = BigInt(0), verbose = false) {
  const log = (...args) => {
    if (verbose) console.log("[x402-solana]", ...args);
  };
  return async (input, init) => {
    const url = typeof input === "string" ? input : input.url;
    log("Making initial request to:", url);
    const response = await fetchFn(input, init);
    log("Initial response status:", response.status);
    if (response.status !== 402) {
      return response;
    }
    log("Got 402, parsing payment requirements...");
    let paymentRequired;
    let protocolVersion;
    const paymentRequiredHeader = response.headers.get("PAYMENT-REQUIRED");
    if (paymentRequiredHeader) {
      log("Found PAYMENT-REQUIRED header (v2 protocol)");
      paymentRequired = decodePaymentRequiredHeader(paymentRequiredHeader);
      protocolVersion = 2;
    } else {
      log("No PAYMENT-REQUIRED header, falling back to body (v1 protocol)");
      const rawResponse = await response.json();
      paymentRequired = rawResponse;
      protocolVersion = 1;
    }
    log("Payment requirements:", JSON.stringify(paymentRequired, null, 2));
    log("Protocol version:", protocolVersion);
    const parsedPaymentRequirements = paymentRequired.accepts || [];
    const selectedRequirements = parsedPaymentRequirements.find(
      (req) => req.scheme === "exact" && isSolanaNetwork(req.network)
    );
    if (!selectedRequirements) {
      console.error(
        "\u274C No suitable Solana payment requirements found. Available networks:",
        parsedPaymentRequirements.map((req) => req.network)
      );
      throw new Error("No suitable Solana payment requirements found");
    }
    const paymentAmount = BigInt(
      selectedRequirements.amount || selectedRequirements.maxAmountRequired || "0"
    );
    if (maxValue > BigInt(0) && paymentAmount > maxValue) {
      throw new Error("Payment amount exceeds maximum allowed");
    }
    const resourceUrl = typeof input === "string" ? input : input.url;
    log("Creating signed transaction...");
    const signedTransaction = await createSolanaPaymentTransaction(
      wallet,
      selectedRequirements,
      rpcUrl
    );
    log("Transaction signed successfully");
    let paymentHeader;
    let headerName;
    if (protocolVersion === 2) {
      paymentHeader = createPaymentPayload(
        signedTransaction,
        selectedRequirements,
        resourceUrl
      );
      headerName = "PAYMENT-SIGNATURE";
    } else {
      paymentHeader = createPaymentPayloadV1(
        signedTransaction,
        selectedRequirements
      );
      headerName = "X-PAYMENT";
    }
    log("Payment header created, length:", paymentHeader.length);
    log("Using header:", headerName);
    const newInit = {
      ...init,
      headers: {
        ...init?.headers || {},
        [headerName]: paymentHeader
      }
    };
    log(`Retrying request with ${headerName} header...`);
    const retryResponse = await fetchFn(input, newInit);
    log("Retry response status:", retryResponse.status);
    return retryResponse;
  };
}

// src/client/index.ts
var X402Client = class {
  paymentFetch;
  constructor(config) {
    const rpcUrl = config.rpcUrl || getDefaultRpcUrl(config.network);
    const fetchFn = config.customFetch || globalThis.fetch.bind(globalThis);
    this.paymentFetch = createPaymentFetch(
      fetchFn,
      config.wallet,
      rpcUrl,
      config.amount || BigInt(0),
      config.verbose || false
    );
  }
  /**
   * Make a fetch request with automatic x402 payment handling.
   * If a customFetch was provided in the config, it will be used for all requests.
   */
  async fetch(input, init) {
    return this.paymentFetch(input, init);
  }
};
function createX402Client(config) {
  return new X402Client(config);
}
var FacilitatorClient = class {
  facilitatorUrl;
  apiKeyId;
  apiKeySecret;
  constructor(config) {
    this.facilitatorUrl = config.url;
    this.apiKeyId = config.apiKeyId;
    this.apiKeySecret = config.apiKeySecret;
  }
  /**
   * Build auth headers when API keys are configured.
   * Returns an empty object when no keys are present.
   */
  async getAuthHeaders() {
    if (!this.apiKeyId || !this.apiKeySecret) {
      return {};
    }
    const jwt = await getOrGenerateJwt({
      apiKeyId: this.apiKeyId,
      apiKeySecret: this.apiKeySecret
    });
    return { Authorization: `Bearer ${jwt}` };
  }
  /**
   * Get supported payment kinds from facilitator
   */
  async getSupported() {
    const authHeaders = await this.getAuthHeaders();
    const response = await fetch(`${this.facilitatorUrl}/supported`, {
      headers: { ...authHeaders }
    });
    if (!response.ok) {
      throw new Error(`Facilitator /supported returned ${response.status}`);
    }
    return response.json();
  }
  /**
   * Get fee payer address for a Solana network
   * @param network - Network in any format (simple or CAIP-2)
   */
  async getFeePayer(network) {
    const supportedData = await this.getSupported();
    const networkSupport = supportedData.kinds?.find(
      (kind) => kind.scheme === "exact" && isSolanaNetwork(kind.network) && isSolanaNetwork(network) && // Match if both are same network type (mainnet or devnet)
      (kind.network.includes("devnet") === network.includes("devnet") || kind.network === network)
    );
    if (!networkSupport?.extra?.feePayer) {
      throw new Error(
        `Facilitator does not support network "${network}" with scheme "exact" or feePayer not provided`
      );
    }
    return networkSupport.extra.feePayer;
  }
  /**
   * Verify payment with facilitator
   * @returns VerifyResponse with isValid and optional invalidReason
   */
  async verifyPayment(paymentHeader, paymentRequirements) {
    try {
      const paymentPayload = JSON.parse(
        Buffer.from(paymentHeader, "base64").toString("utf8")
      );
      const verifyPayload = {
        paymentPayload,
        paymentRequirements
      };
      const authHeaders = await this.getAuthHeaders();
      const response = await fetch(`${this.facilitatorUrl}/verify`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...authHeaders
        },
        body: JSON.stringify(verifyPayload)
      });
      if (!response.ok) {
        const errorBody = await response.text();
        console.error(`Facilitator /verify returned ${response.status}:`, errorBody);
        return {
          isValid: false,
          invalidReason: "unexpected_verify_error"
        };
      }
      const facilitatorResponse = await response.json();
      return facilitatorResponse;
    } catch (error) {
      console.error("Payment verification failed:", error);
      return {
        isValid: false,
        invalidReason: "unexpected_verify_error"
      };
    }
  }
  /**
   * Settle payment with facilitator
   * @returns SettleResponse with success status and optional errorReason
   */
  async settlePayment(paymentHeader, paymentRequirements) {
    try {
      const paymentPayload = JSON.parse(
        Buffer.from(paymentHeader, "base64").toString("utf8")
      );
      const settlePayload = {
        paymentPayload,
        paymentRequirements
      };
      const authHeaders = await this.getAuthHeaders();
      const response = await fetch(`${this.facilitatorUrl}/settle`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...authHeaders
        },
        body: JSON.stringify(settlePayload)
      });
      if (!response.ok) {
        const errorBody = await response.text();
        console.error(`Facilitator /settle returned ${response.status}:`, errorBody);
        return {
          success: false,
          errorReason: "unexpected_settle_error",
          transaction: "",
          network: paymentRequirements.network
        };
      }
      const facilitatorResponse = await response.json();
      return facilitatorResponse;
    } catch (error) {
      console.error("Payment settlement failed:", error);
      return {
        success: false,
        errorReason: "unexpected_settle_error",
        transaction: "",
        network: paymentRequirements.network
      };
    }
  }
};

// src/server/payment-handler.ts
var X402PaymentHandler = class {
  facilitatorClient;
  config;
  constructor(config) {
    const defaultToken = config.defaultToken || getDefaultTokenAsset(config.network);
    this.config = {
      network: toCAIP2Network(config.network),
      treasuryAddress: config.treasuryAddress,
      facilitatorUrl: config.facilitatorUrl,
      rpcUrl: config.rpcUrl || getDefaultRpcUrl(config.network),
      defaultToken,
      defaultDescription: config.defaultDescription || "Payment required",
      defaultTimeoutSeconds: config.defaultTimeoutSeconds || 300
    };
    this.facilitatorClient = new FacilitatorClient({
      url: config.facilitatorUrl,
      apiKeyId: config.apiKeyId,
      apiKeySecret: config.apiKeySecret
    });
  }
  /**
   * Extract payment header from request headers (v2)
   * Pass in headers object from any framework (Next.js, Express, etc.)
   *
   * Note: v2 uses PAYMENT-SIGNATURE header (v1 used X-PAYMENT)
   */
  extractPayment(headers) {
    if (headers instanceof Headers) {
      return headers.get("PAYMENT-SIGNATURE") || headers.get("payment-signature");
    }
    const paymentSignature = headers["PAYMENT-SIGNATURE"] || headers["payment-signature"];
    return Array.isArray(paymentSignature) ? paymentSignature[0] || null : paymentSignature || null;
  }
  /**
   * Create payment requirements for a protected resource
   *
   * @param routeConfig - Route-specific configuration
   * @param resourceUrl - URL of the protected resource
   * @returns Payment requirements object
   */
  async createPaymentRequirements(routeConfig, resourceUrl) {
    const feePayer = await this.facilitatorClient.getFeePayer(this.config.network);
    const paymentRequirements = {
      scheme: "exact",
      network: this.config.network,
      amount: routeConfig.amount,
      payTo: this.config.treasuryAddress,
      maxTimeoutSeconds: routeConfig.maxTimeoutSeconds || this.config.defaultTimeoutSeconds,
      asset: routeConfig.asset.address,
      extra: {
        feePayer,
        description: routeConfig.description || this.config.defaultDescription,
        mimeType: routeConfig.mimeType || "application/json",
        resource: resourceUrl
      }
    };
    return paymentRequirements;
  }
  /**
   * Create a 402 Payment Required response body (v2)
   * Use this with your framework's response method
   *
   * @param requirements - Payment requirements (from createPaymentRequirements)
   * @param resourceUrl - URL of the protected resource
   */
  create402Response(requirements, resourceUrl) {
    return {
      status: 402,
      body: {
        x402Version: 2,
        resource: {
          url: resourceUrl,
          description: requirements.extra?.description || "",
          mimeType: requirements.extra?.mimeType || "application/json"
        },
        accepts: [requirements],
        error: "Payment required"
      }
    };
  }
  /**
   * Verify payment with facilitator
   * @returns VerifyResponse with isValid and optional invalidReason
   */
  async verifyPayment(paymentHeader, paymentRequirements) {
    return this.facilitatorClient.verifyPayment(paymentHeader, paymentRequirements);
  }
  /**
   * Settle payment with facilitator
   * @returns SettleResponse with success status and optional errorReason
   */
  async settlePayment(paymentHeader, paymentRequirements) {
    return this.facilitatorClient.settlePayment(paymentHeader, paymentRequirements);
  }
  /**
   * Get the network in CAIP-2 format
   */
  getNetwork() {
    return this.config.network;
  }
  /**
   * Get the treasury address
   */
  getTreasuryAddress() {
    return this.config.treasuryAddress;
  }
};

export { FacilitatorClient, SOLANA_DEVNET_CAIP2, SOLANA_MAINNET_CAIP2, X402Client, X402PaymentHandler, createPaymentPayload, createPaymentPayloadV1, createX402Client, fromAtomicUnits, getDefaultRpcUrl, getDefaultTokenAsset, getRpcUrlForNetwork, isSolanaDevnet, isSolanaMainnet, isSolanaNetwork, toAtomicUnits, toCAIP2Network, toSimpleNetwork };
