// x402-solana/dist/server/index.js
'use strict';

var facilitator = require('@payai/facilitator');

// src/types/x402-protocol.ts
var SOLANA_MAINNET_CAIP2 = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
var SOLANA_DEVNET_CAIP2 = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1";
function isSolanaNetwork(network) {
  return network === "solana" || network === "solana-devnet" || network.startsWith("solana:");
}
function isSolanaMainnet(network) {
  return network === "solana" || network === SOLANA_MAINNET_CAIP2;
}
function toCAIP2Network(network) {
  switch (network) {
    case "solana":
      return SOLANA_MAINNET_CAIP2;
    case "solana-devnet":
      return SOLANA_DEVNET_CAIP2;
  }
}

// src/utils/helpers.ts
var USDC_MAINNET = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
var USDC_DEVNET = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";
function getDefaultRpcUrl(network) {
  if (isSolanaMainnet(network)) {
    return "https://api.mainnet-beta.solana.com";
  }
  return "https://api.devnet.solana.com";
}
function getDefaultTokenAsset(network) {
  if (isSolanaMainnet(network)) {
    return {
      address: USDC_MAINNET,
      decimals: 6
    };
  }
  return {
    address: USDC_DEVNET,
    decimals: 6
  };
}
var FacilitatorClient = class {
  facilitatorUrl;
  apiKeyId;
  apiKeySecret;
  constructor(config) {
    this.facilitatorUrl = config.url;
    this.apiKeyId = config.apiKeyId;
    this.apiKeySecret = config.apiKeySecret;
  }
  /**
   * Build auth headers when API keys are configured.
   * Returns an empty object when no keys are present.
   */
  async getAuthHeaders() {
    if (!this.apiKeyId || !this.apiKeySecret) {
      return {};
    }
    const jwt = await facilitator.getOrGenerateJwt({
      apiKeyId: this.apiKeyId,
      apiKeySecret: this.apiKeySecret
    });
    return { Authorization: `Bearer ${jwt}` };
  }
  /**
   * Get supported payment kinds from facilitator
   */
  async getSupported() {
    const authHeaders = await this.getAuthHeaders();
    const response = await fetch(`${this.facilitatorUrl}/supported`, {
      headers: { ...authHeaders }
    });
    if (!response.ok) {
      throw new Error(`Facilitator /supported returned ${response.status}`);
    }
    return response.json();
  }
  /**
   * Get fee payer address for a Solana network
   * @param network - Network in any format (simple or CAIP-2)
   */
  async getFeePayer(network) {
    const supportedData = await this.getSupported();
    const networkSupport = supportedData.kinds?.find(
      (kind) => kind.scheme === "exact" && isSolanaNetwork(kind.network) && isSolanaNetwork(network) && // Match if both are same network type (mainnet or devnet)
      (kind.network.includes("devnet") === network.includes("devnet") || kind.network === network)
    );
    if (!networkSupport?.extra?.feePayer) {
      throw new Error(
        `Facilitator does not support network "${network}" with scheme "exact" or feePayer not provided`
      );
    }
    return networkSupport.extra.feePayer;
  }
  /**
   * Verify payment with facilitator
   * @returns VerifyResponse with isValid and optional invalidReason
   */
  async verifyPayment(paymentHeader, paymentRequirements) {
    try {
      const paymentPayload = JSON.parse(
        Buffer.from(paymentHeader, "base64").toString("utf8")
      );
      const verifyPayload = {
        paymentPayload,
        paymentRequirements
      };
      const authHeaders = await this.getAuthHeaders();
      const response = await fetch(`${this.facilitatorUrl}/verify`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...authHeaders
        },
        body: JSON.stringify(verifyPayload)
      });
      if (!response.ok) {
        const errorBody = await response.text();
        console.error(`Facilitator /verify returned ${response.status}:`, errorBody);
        return {
          isValid: false,
          invalidReason: "unexpected_verify_error"
        };
      }
      const facilitatorResponse = await response.json();
      return facilitatorResponse;
    } catch (error) {
      console.error("Payment verification failed:", error);
      return {
        isValid: false,
        invalidReason: "unexpected_verify_error"
      };
    }
  }
  /**
   * Settle payment with facilitator
   * @returns SettleResponse with success status and optional errorReason
   */
  async settlePayment(paymentHeader, paymentRequirements) {
    try {
      const paymentPayload = JSON.parse(
        Buffer.from(paymentHeader, "base64").toString("utf8")
      );
      const settlePayload = {
        paymentPayload,
        paymentRequirements
      };
      const authHeaders = await this.getAuthHeaders();
      const response = await fetch(`${this.facilitatorUrl}/settle`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...authHeaders
        },
        body: JSON.stringify(settlePayload)
      });
      if (!response.ok) {
        const errorBody = await response.text();
        console.error(`Facilitator /settle returned ${response.status}:`, errorBody);
        return {
          success: false,
          errorReason: "unexpected_settle_error",
          transaction: "",
          network: paymentRequirements.network
        };
      }
      const facilitatorResponse = await response.json();
      return facilitatorResponse;
    } catch (error) {
      console.error("Payment settlement failed:", error);
      return {
        success: false,
        errorReason: "unexpected_settle_error",
        transaction: "",
        network: paymentRequirements.network
      };
    }
  }
};

// src/server/payment-handler.ts
var X402PaymentHandler = class {
  facilitatorClient;
  config;
  constructor(config) {
    const defaultToken = config.defaultToken || getDefaultTokenAsset(config.network);
    this.config = {
      network: toCAIP2Network(config.network),
      treasuryAddress: config.treasuryAddress,
      facilitatorUrl: config.facilitatorUrl,
      rpcUrl: config.rpcUrl || getDefaultRpcUrl(config.network),
      defaultToken,
      defaultDescription: config.defaultDescription || "Payment required",
      defaultTimeoutSeconds: config.defaultTimeoutSeconds || 300
    };
    this.facilitatorClient = new FacilitatorClient({
      url: config.facilitatorUrl,
      apiKeyId: config.apiKeyId,
      apiKeySecret: config.apiKeySecret
    });
  }
  /**
   * Extract payment header from request headers (v2)
   * Pass in headers object from any framework (Next.js, Express, etc.)
   *
   * Note: v2 uses PAYMENT-SIGNATURE header (v1 used X-PAYMENT)
   */
  extractPayment(headers) {
    if (headers instanceof Headers) {
      return headers.get("PAYMENT-SIGNATURE") || headers.get("payment-signature");
    }
    const paymentSignature = headers["PAYMENT-SIGNATURE"] || headers["payment-signature"];
    return Array.isArray(paymentSignature) ? paymentSignature[0] || null : paymentSignature || null;
  }
  /**
   * Create payment requirements for a protected resource
   *
   * @param routeConfig - Route-specific configuration
   * @param resourceUrl - URL of the protected resource
   * @returns Payment requirements object
   */
  async createPaymentRequirements(routeConfig, resourceUrl) {
    const feePayer = await this.facilitatorClient.getFeePayer(this.config.network);
    const paymentRequirements = {
      scheme: "exact",
      network: this.config.network,
      amount: routeConfig.amount,
      payTo: this.config.treasuryAddress,
      maxTimeoutSeconds: routeConfig.maxTimeoutSeconds || this.config.defaultTimeoutSeconds,
      asset: routeConfig.asset.address,
      extra: {
        feePayer,
        description: routeConfig.description || this.config.defaultDescription,
        mimeType: routeConfig.mimeType || "application/json",
        resource: resourceUrl
      }
    };
    return paymentRequirements;
  }
  /**
   * Create a 402 Payment Required response body (v2)
   * Use this with your framework's response method
   *
   * @param requirements - Payment requirements (from createPaymentRequirements)
   * @param resourceUrl - URL of the protected resource
   */
  create402Response(requirements, resourceUrl) {
    return {
      status: 402,
      body: {
        x402Version: 2,
        resource: {
          url: resourceUrl,
          description: requirements.extra?.description || "",
          mimeType: requirements.extra?.mimeType || "application/json"
        },
        accepts: [requirements],
        error: "Payment required"
      }
    };
  }
  /**
   * Verify payment with facilitator
   * @returns VerifyResponse with isValid and optional invalidReason
   */
  async verifyPayment(paymentHeader, paymentRequirements) {
    return this.facilitatorClient.verifyPayment(paymentHeader, paymentRequirements);
  }
  /**
   * Settle payment with facilitator
   * @returns SettleResponse with success status and optional errorReason
   */
  async settlePayment(paymentHeader, paymentRequirements) {
    return this.facilitatorClient.settlePayment(paymentHeader, paymentRequirements);
  }
  /**
   * Get the network in CAIP-2 format
   */
  getNetwork() {
    return this.config.network;
  }
  /**
   * Get the treasury address
   */
  getTreasuryAddress() {
    return this.config.treasuryAddress;
  }
};

exports.FacilitatorClient = FacilitatorClient;
exports.X402PaymentHandler = X402PaymentHandler;
