// x402-solana/dist/client/index.mjs
import { safeBase64Decode } from '@payai/x402/utils';
import { Connection, PublicKey, ComputeBudgetProgram, TransactionMessage, VersionedTransaction } from '@solana/web3.js';
import { TOKEN_2022_PROGRAM_ID, TOKEN_PROGRAM_ID, getMint, getAssociatedTokenAddress, createTransferCheckedInstruction } from '@solana/spl-token';

// src/types/x402-protocol.ts
var SOLANA_MAINNET_CAIP2 = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
function isSolanaNetwork(network) {
  return network === "solana" || network === "solana-devnet" || network.startsWith("solana:");
}
function isSolanaMainnet(network) {
  return network === "solana" || network === SOLANA_MAINNET_CAIP2;
}

// src/utils/helpers.ts
function createPaymentPayload(transaction, paymentRequirements, resourceUrl) {
  const base64Transaction = Buffer.from(transaction.serialize()).toString(
    "base64"
  );
  const paymentPayload = {
    x402Version: 2,
    resource: {
      url: resourceUrl,
      description: paymentRequirements.extra?.description || "",
      mimeType: paymentRequirements.extra?.mimeType || "application/json"
    },
    accepted: paymentRequirements,
    payload: {
      transaction: base64Transaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString("base64");
  return paymentHeader;
}
function createPaymentPayloadV1(transaction, paymentRequirements) {
  const base64Transaction = Buffer.from(transaction.serialize()).toString("base64");
  const paymentPayload = {
    x402Version: 1,
    scheme: paymentRequirements.scheme,
    network: paymentRequirements.network,
    payload: {
      transaction: base64Transaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString(
    "base64"
  );
  return paymentHeader;
}
function getDefaultRpcUrl(network) {
  if (isSolanaMainnet(network)) {
    return "https://api.mainnet-beta.solana.com";
  }
  return "https://api.devnet.solana.com";
}
var DEFAULT_COMPUTE_UNIT_LIMIT = 2e4;
var DEFAULT_COMPUTE_UNIT_PRICE_MICROLAMPORTS = 1;
async function createSolanaPaymentTransaction(wallet, paymentRequirements, rpcUrl) {
  const connection = new Connection(rpcUrl, "confirmed");
  const feePayer = paymentRequirements.extra?.feePayer;
  if (!feePayer) {
    throw new Error(
      "Missing facilitator feePayer in payment requirements (extra.feePayer)."
    );
  }
  const feePayerPubkey = new PublicKey(feePayer);
  const walletAddress = wallet?.publicKey?.toString() || wallet?.address;
  if (!walletAddress) {
    throw new Error("Missing connected Solana wallet address or publicKey");
  }
  const userPubkey = new PublicKey(walletAddress);
  if (!paymentRequirements.payTo) {
    throw new Error("Missing payTo in payment requirements");
  }
  const destination = new PublicKey(paymentRequirements.payTo);
  if (!paymentRequirements.asset) {
    throw new Error("Missing token mint for SPL transfer");
  }
  const mintPubkey = new PublicKey(paymentRequirements.asset);
  const instructions = [];
  instructions.push(
    ComputeBudgetProgram.setComputeUnitLimit({
      units: DEFAULT_COMPUTE_UNIT_LIMIT
    })
  );
  instructions.push(
    ComputeBudgetProgram.setComputeUnitPrice({
      microLamports: DEFAULT_COMPUTE_UNIT_PRICE_MICROLAMPORTS
    })
  );
  const mintInfo = await connection.getAccountInfo(mintPubkey, "confirmed");
  const programId = mintInfo?.owner?.toBase58() === TOKEN_2022_PROGRAM_ID.toBase58() ? TOKEN_2022_PROGRAM_ID : TOKEN_PROGRAM_ID;
  const mint = await getMint(connection, mintPubkey, void 0, programId);
  const sourceAta = await getAssociatedTokenAddress(
    mintPubkey,
    userPubkey,
    false,
    programId
  );
  const destinationAta = await getAssociatedTokenAddress(
    mintPubkey,
    destination,
    false,
    programId
  );
  const sourceAtaInfo = await connection.getAccountInfo(sourceAta, "confirmed");
  if (!sourceAtaInfo) {
    throw new Error(
      `User does not have an Associated Token Account for ${paymentRequirements.asset}. Please create one first or ensure you have the required token.`
    );
  }
  const destAtaInfo = await connection.getAccountInfo(
    destinationAta,
    "confirmed"
  );
  if (!destAtaInfo) {
    throw new Error(
      `Destination does not have an Associated Token Account for ${paymentRequirements.asset}. The receiver must create their token account before receiving payments.`
    );
  }
  const amountStr = paymentRequirements.amount || paymentRequirements.maxAmountRequired;
  if (!amountStr) {
    throw new Error("Missing amount in payment requirements");
  }
  const amount = BigInt(amountStr);
  instructions.push(
    createTransferCheckedInstruction(
      sourceAta,
      mintPubkey,
      destinationAta,
      userPubkey,
      amount,
      mint.decimals,
      [],
      programId
    )
  );
  const { blockhash } = await connection.getLatestBlockhash("confirmed");
  const message = new TransactionMessage({
    payerKey: feePayerPubkey,
    recentBlockhash: blockhash,
    instructions
  }).compileToV0Message();
  const transaction = new VersionedTransaction(message);
  if (typeof wallet?.signTransaction !== "function") {
    throw new Error("Connected wallet does not support signTransaction");
  }
  const userSignedTx = await wallet.signTransaction(transaction);
  return userSignedTx;
}

// src/client/payment-interceptor.ts
function decodePaymentRequiredHeader(header) {
  const decoded = safeBase64Decode(header);
  return JSON.parse(decoded);
}
function createPaymentFetch(fetchFn, wallet, rpcUrl, maxValue = BigInt(0), verbose = false) {
  const log = (...args) => {
    if (verbose) console.log("[x402-solana]", ...args);
  };
  return async (input, init) => {
    const url = typeof input === "string" ? input : input.url;
    log("Making initial request to:", url);
    const response = await fetchFn(input, init);
    log("Initial response status:", response.status);
    if (response.status !== 402) {
      return response;
    }
    log("Got 402, parsing payment requirements...");
    let paymentRequired;
    let protocolVersion;
    const paymentRequiredHeader = response.headers.get("PAYMENT-REQUIRED");
    if (paymentRequiredHeader) {
      log("Found PAYMENT-REQUIRED header (v2 protocol)");
      paymentRequired = decodePaymentRequiredHeader(paymentRequiredHeader);
      protocolVersion = 2;
    } else {
      log("No PAYMENT-REQUIRED header, falling back to body (v1 protocol)");
      const rawResponse = await response.json();
      paymentRequired = rawResponse;
      protocolVersion = 1;
    }
    log("Payment requirements:", JSON.stringify(paymentRequired, null, 2));
    log("Protocol version:", protocolVersion);
    const parsedPaymentRequirements = paymentRequired.accepts || [];
    const selectedRequirements = parsedPaymentRequirements.find(
      (req) => req.scheme === "exact" && isSolanaNetwork(req.network)
    );
    if (!selectedRequirements) {
      console.error(
        "\u274C No suitable Solana payment requirements found. Available networks:",
        parsedPaymentRequirements.map((req) => req.network)
      );
      throw new Error("No suitable Solana payment requirements found");
    }
    const paymentAmount = BigInt(
      selectedRequirements.amount || selectedRequirements.maxAmountRequired || "0"
    );
    if (maxValue > BigInt(0) && paymentAmount > maxValue) {
      throw new Error("Payment amount exceeds maximum allowed");
    }
    const resourceUrl = typeof input === "string" ? input : input.url;
    log("Creating signed transaction...");
    const signedTransaction = await createSolanaPaymentTransaction(
      wallet,
      selectedRequirements,
      rpcUrl
    );
    log("Transaction signed successfully");
    let paymentHeader;
    let headerName;
    if (protocolVersion === 2) {
      paymentHeader = createPaymentPayload(
        signedTransaction,
        selectedRequirements,
        resourceUrl
      );
      headerName = "PAYMENT-SIGNATURE";
    } else {
      paymentHeader = createPaymentPayloadV1(
        signedTransaction,
        selectedRequirements
      );
      headerName = "X-PAYMENT";
    }
    log("Payment header created, length:", paymentHeader.length);
    log("Using header:", headerName);
    const newInit = {
      ...init,
      headers: {
        ...init?.headers || {},
        [headerName]: paymentHeader
      }
    };
    log(`Retrying request with ${headerName} header...`);
    const retryResponse = await fetchFn(input, newInit);
    log("Retry response status:", retryResponse.status);
    return retryResponse;
  };
}

// src/client/index.ts
var X402Client = class {
  paymentFetch;
  constructor(config) {
    const rpcUrl = config.rpcUrl || getDefaultRpcUrl(config.network);
    const fetchFn = config.customFetch || globalThis.fetch.bind(globalThis);
    this.paymentFetch = createPaymentFetch(
      fetchFn,
      config.wallet,
      rpcUrl,
      config.amount || BigInt(0),
      config.verbose || false
    );
  }
  /**
   * Make a fetch request with automatic x402 payment handling.
   * If a customFetch was provided in the config, it will be used for all requests.
   */
  async fetch(input, init) {
    return this.paymentFetch(input, init);
  }
};
function createX402Client(config) {
  return new X402Client(config);
}

export { X402Client, createX402Client };
