// @payai/facilitator/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  clearJwtCache: () => clearJwtCache,
  createFacilitatorConfig: () => createFacilitatorConfig,
  createPayAIAuthHeaders: () => createPayAIAuthHeaders,
  facilitator: () => facilitator,
  getOrGenerateJwt: () => getOrGenerateJwt
});
module.exports = __toCommonJS(src_exports);

// src/payaiJwt.ts
function base64UrlEncode(data) {
  let binary = "";
  for (let i = 0; i < data.byteLength; i++) {
    binary += String.fromCharCode(data[i]);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
function base64UrlEncodeString(str) {
  return base64UrlEncode(new TextEncoder().encode(str));
}
function base64ToUint8Array(base64) {
  const standardized = base64.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(standardized);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}
function normalizeApiKeySecret(apiKeySecret) {
  const trimmed = apiKeySecret.trim();
  return trimmed.startsWith("payai_sk_") ? trimmed.slice("payai_sk_".length) : trimmed;
}
function generateUUID() {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = Math.random() * 16 | 0;
    const v = c === "x" ? r : r & 3 | 8;
    return v.toString(16);
  });
}
async function generatePayAIApiKeyJwt(options) {
  const { apiKeyId, apiKeySecret, expiresIn = 120, issuer = "payai-merchant" } = options;
  if (!apiKeyId || !apiKeySecret) {
    throw new Error("apiKeyId and apiKeySecret are required");
  }
  const now = Math.floor(Date.now() / 1e3);
  const header = { alg: "EdDSA", typ: "JWT", kid: apiKeyId };
  const payload = {
    sub: apiKeyId,
    iss: issuer,
    iat: now,
    exp: now + expiresIn,
    jti: generateUUID()
  };
  const headerB64 = base64UrlEncodeString(JSON.stringify(header));
  const payloadB64 = base64UrlEncodeString(JSON.stringify(payload));
  const message = `${headerB64}.${payloadB64}`;
  try {
    const keyBytes = base64ToUint8Array(normalizeApiKeySecret(apiKeySecret));
    const privateKey = await crypto.subtle.importKey(
      "pkcs8",
      keyBytes.buffer,
      { name: "Ed25519" },
      false,
      ["sign"]
    );
    const signature = await crypto.subtle.sign(
      "Ed25519",
      privateKey,
      new TextEncoder().encode(message).buffer
    );
    return `${message}.${base64UrlEncode(new Uint8Array(signature))}`;
  } catch (error) {
    const msg = error instanceof Error ? error.message : "Unknown error";
    throw new Error(
      `Failed to sign JWT: ${msg}. Ensure apiKeySecret is a valid base64-encoded Ed25519 private key in PKCS#8 format.`
    );
  }
}
var jwtCache = /* @__PURE__ */ new Map();
var REFRESH_BUFFER_SECONDS = 30;
async function getOrGenerateJwt(options) {
  const now = Math.floor(Date.now() / 1e3);
  const cached = jwtCache.get(options.apiKeyId);
  if (cached && cached.expiresAt > now + REFRESH_BUFFER_SECONDS) {
    return cached.token;
  }
  const expiresIn = options.expiresIn ?? 120;
  const token = await generatePayAIApiKeyJwt(options);
  jwtCache.set(options.apiKeyId, { token, expiresAt: now + expiresIn });
  return token;
}
function clearJwtCache(apiKeyId) {
  if (apiKeyId) {
    jwtCache.delete(apiKeyId);
  } else {
    jwtCache.clear();
  }
}
function createPayAIAuthHeaders(apiKeyId, apiKeySecret) {
  return async () => {
    const jwt = await getOrGenerateJwt({ apiKeyId, apiKeySecret });
    const authHeader = { Authorization: `Bearer ${jwt}` };
    return {
      verify: authHeader,
      settle: authHeader,
      supported: authHeader
    };
  };
}

// src/index.ts
var PAYAI_FACILITATOR_URL = "https://facilitator.payai.network";
function createFacilitatorConfig(apiKeyId, apiKeySecret) {
  const keyId = apiKeyId ?? process.env.PAYAI_API_KEY_ID;
  const keySecret = apiKeySecret ?? process.env.PAYAI_API_KEY_SECRET;
  const config = { url: PAYAI_FACILITATOR_URL };
  if (keyId && keySecret) {
    config.createAuthHeaders = createPayAIAuthHeaders(keyId, keySecret);
  }
  return config;
}
var facilitator = {
  url: PAYAI_FACILITATOR_URL,
  /**
   * Read credentials from env at request-time.
   *
   * This avoids evaluating `process.env` during module initialization, which makes
   * the default export safe to import in non-Node runtimes (keys can still be passed
   * explicitly via `createFacilitatorConfig()`).
   *
   * @returns auth headers per facilitator endpoint, or empty objects for free tier
   */
  createAuthHeaders: async () => {
    const keyId = process.env.PAYAI_API_KEY_ID;
    const keySecret = process.env.PAYAI_API_KEY_SECRET;
    if (!keyId || !keySecret) {
      return { verify: {}, settle: {}, supported: {} };
    }
    return createPayAIAuthHeaders(keyId, keySecret)();
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  clearJwtCache,
  createFacilitatorConfig,
  createPayAIAuthHeaders,
  facilitator,
  getOrGenerateJwt
});
